clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,740 @@
+//===- SYCLOffloadWrapper.cpp -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// This is a part of wrapping functionality related to SYCL part.
+// All implementation details were inspired by previous implementation in
+// clang-offload-wrapper tool.
+// OffloadWrapper accepts device images, corresponding entries, properties,
+// and additional optional options. Then it wraps this into a LLVM IR
+// Module accordingly to the expected format listed in
+//  sycl/include/sycl/detail/pi.h
+//===----------------------------------------------------------------------===//
+
+#include "SYCLOffloadWrapper.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Frontend/Offloading/Utility.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Type.h"
+#include "llvm/SYCLLowerIR/SYCLUtils.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/LineIterator.h"
+#include "llvm/Support/PropertySetIO.h"
+#include "llvm/Transforms/Utils/ModuleUtils.h"
+#include <memory>
+#include <string>
+#include <utility>
+
+using namespace llvm;
+using namespace llvm::util;
+
+namespace {
+
+/// Note: Returned values are a part of ABI. If you want to change them
+/// then coordinate it with SYCL Runtime.
+int8_t binaryImageFormatToInt8(SYCLBinaryImageFormat Format) {
+  switch (Format) {
+  case SYCLBinaryImageFormat::BIF_None:
+    return 0;
+  case SYCLBinaryImageFormat::BIF_Native:
+    return 1;
+  case SYCLBinaryImageFormat::BIF_SPIRV:
+    return 2;
+  case SYCLBinaryImageFormat::BIF_LLVMBC:
+    return 3;
+  default:
+    llvm_unreachable("unexpected SYCLBinaryImageFormat");
+  }
+}
+
+/// Wrapper helper class that creates all LLVM IRs wrapping given images.
```

Comment:
We might need to add a separate document to describe the wrapped file structure for SYCL. It can be done in a subsequent PR.

Thanks